### PR TITLE
Add support for SM4-XTS/CTS encryption

### PR DESCRIPTION
**Pull request by uudiin**
This was added in Linux v6.2-rc1.

### CODE REVIEW

File: test.py [hunk 1]
```diff
@@ -342,6 +342,38 @@ def test_set_get_policy_aes_128_cbc(directory):
                      filenames="AES-128-CTS")
 
 
+def test_set_get_policy_sm4(directory):
+    """Tests getting and setting an encryption policy that uses SM4-XTS
+    contents encryption and SM4-CTS filenames encryption.  Note that the
+    kernel doesn't yet support SM4-XTS contents encryption and SM4-CTS
+    filenames encryption, so that is not tested."""
```
Comment (ebiggers) [inline test.py:349]: The second sentence of the above comment makes no sense.

File: fscrypt_uapi.h [hunk 2]
```diff
@@ -26,6 +26,8 @@
 #define FSCRYPT_MODE_AES_256_CTS		4
 #define FSCRYPT_MODE_AES_128_CBC		5
 #define FSCRYPT_MODE_AES_128_CTS		6
+#define FSCRYPT_MODE_SM4_XTS			7
+#define FSCRYPT_MODE_SM4_CTS			8
```
Comment (ebiggers) [inline fscrypt_uapi.h:30]: Please sync this file from upstream.  There was not the only change to it.

File: test.py [hunk 3]
```diff
@@ -342,6 +342,38 @@ def test_set_get_policy_aes_128_cbc(directory):
...
+                              "--filenames=SM4-CTS")
+    except SystemError as e:
+        # Old kernel that doesn't know about SM4-XTS and SM4-CTS
+        assert "invalid encryption options provided" in str(e)
+        pytest.skip("Kernel doesn't support SM4-XTS and SM4-CTS encryption")
+    except OSError as e:
+        assert "Package not installed" in str(e)
+        pytest.skip("Kernel doesn't support SM4-XTS and SM4-CTS encryption")
+
+    for padding in [4, 16, 32, None]:
```
Comment (ebiggers) [inline test.py:363]: Maybe leave out the non-default filename padding tests, for simplicity.  Non-default filenames padding is already tested with other modes.  Testing it again with SM4 adds very little value, IMO.

**Review by ebiggers — CHANGES_REQUESTED** [review]
Why does the title of the pull request mention HCTR2?  This has nothing to do with HCTR2.

**Comment by uudiin** [comment]
> Why does the title of the pull request mention HCTR2? This has nothing to do with HCTR2.

It's a pity that SM4 failed to enable HCTR2 mode, I have deleted the relevant description.

**Pull request merged**